File: bench/strbuild.rs
// Rust twin of strbuild.ts: the same join tree over immutable strings, where
// every concatenation allocates a fresh String (`[a, b].concat()`) and the
// inputs are dropped, matching strbuild_naive.c. (The idiomatic
// `String::push_str` would grow one buffer in place instead and is O(n);
// docs/BENCHMARKS.md reports it separately.)
const N: i32 = 131072; // bench:n
const FANOUT: i32 = 32;

fn piece(i: i32) -> String {
    format!("{},", i)
}

fn join(lo: i32, hi: i32) -> String {
    let count = hi - lo;
    if count <= FANOUT {
        let mut s = String::new();
        for i in lo..hi {
            s = [s.as_str(), piece(i).as_str()].concat();
        }
        return s;
    }
    let step = (count + FANOUT - 1) / FANOUT;
    let mut s = String::new();
    let mut start = lo;
    while start < hi {
        let end = if start + step < hi { start + step } else { hi };
        s = [s.as_str(), join(start, end).as_str()].concat();
        start += step;
    }
    s
}

fn main() {
    let s = join(0, N);
    println!("{}", s.len());
}

File: bench/strbuild.go
// Go twin of strbuild.ts: the same 32-way join tree over immutable strings.
// Go strings are immutable, so `s = s + piece(i)` allocates a fresh one per
// concatenation and drops its inputs, the same work as strbuild_naive.c and
// the Rust twin; the garbage collector, not an arena, reclaims them.
// (`strings.Builder` would grow one buffer in place instead and is O(n).)
// `strconv.Itoa(i) + ","` is the direct transliteration of the `${i},`
// template literal; `fmt.Sprintf` would time the formatter instead.
package main

import (
	"fmt"
	"strconv"
)

const N = 131072 // bench:n
const FANOUT = 32

func piece(i int32) string {
	return strconv.Itoa(int(i)) + ","
}

func join(lo int32, hi int32) string {
	count := hi - lo
	if count <= FANOUT {
		s := ""
		for i := lo; i < hi; i++ {
			s = s + piece(i)
		}
		return s
	}
	step := (count + FANOUT - 1) / FANOUT
	s := ""
	for start := lo; start < hi; start += step {
		end := hi
		if start+step < hi {
			end = start + step
		}
		s = s + join(start, end)
	}
	return s
}

func main() {
	s := join(0, N)
	fmt.Println(len(s))
}
